FRAuth/FRAuth/AppIntegrity/FRAppAttestDomainModal.swift

```diff
@@ -0,0 +1,136 @@
+    /// - Parameter bundleIdentifier: BundleId of the application
+    /// - encoder encoder: Encoder to endode the Data
+    public init(service: FRAppAttestService = FRAppAttestServiceImpl(),
+                bundleIdentifier: String? = Bundle.main.bundleIdentifier,
+                encoder: JSONEncoder = JSONEncoder()) {
+        self.service = service
+        self.bundleIdentifier = bundleIdentifier
+        self.encoder = encoder
+    }
+    
+    /// Handle attestation and assertion
+    /// - Parameter challenge: Challenge Received from server
+    /// - Returns: FRAppIntegrityKeys for attestation and assertion
+    public func attest(challenge: String) async throws -> FRAppIntegrityKeys {
+        guard let challengeUtf8 = challenge.data(using: .utf8), !challengeUtf8.isEmpty else {
+            throw FRAppIntegrityFailure.invalidChallenge
+        }
+        guard let bundleIdentifier = bundleIdentifier, !bundleIdentifier.isEmpty else {
+            throw FRAppIntegrityFailure.invalidBundleIdentifier
+        }
+        let userClientData = [challengeKey: challenge,
+                               bundleIdKey: bundleIdentifier]
+        
+        guard let jsonData = try? encoder.encode(userClientData) else {
+            throw FRAppIntegrityFailure.invalidClientData
+        }
+        
+        if !service.isSupported() {
+            throw FRAppIntegrityFailure.featureUnsupported
+        }
+        
+        do {
+            let keyIdentifier = try await service.generateKey()
+            let attest = try await service.attest(keyIdentifier: keyIdentifier, clientDataHash: Data(SHA256.hash(data: challengeUtf8)))
+            let assert = try await service.generateAssertion(keyIdentifier: keyIdentifier, clientDataHash: Data(SHA256.hash(data: jsonData)))
+            return FRAppIntegrityKeys(attestKey: attest.base64EncodedString(),
+                                      assertKey: assert.base64EncodedString(),
+                                      keyIdentifier: keyIdentifier,
+                                      clientDataHash: jsonData.base64EncodedString())
+        }
+        catch let error as DCError {
+            throw FRAppIntegrityFailure.error(code: error.errorCode)
+        }
+        catch {
+            throw error
+        }
+    }
+    
+}
+
+/// Results of AppIntegrity Failures
+/// We need to return some of failures for iOS12, iOS13 devices as well.
+public enum FRAppIntegrityFailure: String, Error {
+    case unknownSystemFailure
+    case featureUnsupported
+    case invalidInput
+    case invalidKey
+    case serverUnavailable
+    case invalidChallenge
+    case invalidBundleIdentifier
+    case invalidClientData
+    case unknownError
+    
+
+    static func clientError(failure: FRAppIntegrityFailure) -> String {
+        switch failure {
+         case .featureUnsupported:
+            return "Unsupported"
+        default:
+            return "ClientDeviceErrors"
```

Comment:
I suggest we centralize the strings for errors or outcomes. Either in a utility class (we might have one already) or at least top of the file, on a let

Reply:
taken care , i abstracted all the errors in to his own enum.